### ISSUE TITLE
Persistent Tensor deletion fails when tensor was moved

### ISSUE THREAD

**Issue by yaroslavvb** [tensorflow/tensorflow#2645]
TLDR; handle movers register Python SessionHandle instead of string session handle, which causes `InternalError: Unable to get element from the feed.` when garbage collection is triggered on moved persistent tensors.

The solution is to replace `self._register_dead_handle(handle)` with `self._register_dead_handle(handle.handle)` in `session.py`

## Discussion 
There are two types of handle objects, Python TensorHandle and "native" string session handle. The conversion from Python `TensorHandle` object to native handle `string` object is done in `session.py` and `session_ops.py` in one of those 3 ways.

- `self._handle = compat.as_str_any(handle)`
- `str(handle)`
- `handle.handle`

The "str" conversion is not very Python since `__str__` is supposed to be used for display/debugging purposes. Some suggestions to make it easier to avoid bugs like above:
1. Annotate which kind of handle object is being dealt with. Perhaps `py_handle` variable should refer to Python TensorHandle objects and `handle` refer to native "string" handle objects
2. Only use one kind of conversion method, perhaps as `py_handle.handle`
3. `__str__` method of TensorHandle could be `return `"TensorHandle(%s)"%(self.handle)` for easier debugging

## Test case
```
class EnvTest(tf.test.TestCase):

  def testHandleDeletion(self):
    dtype = tf.float32
    # soft-placement to work around #2587
    config = tf.ConfigProto(log_device_placement=True,
                            allow_soft_placement=True)
    sess = tf.Session(config=config)

    # initial values live on CPU
    with tf.device("/cpu:0"):
      one = tf.constant(1, dtype=dtype)
      one_handle = sess.run(tf.get_session_handle(one))
      x = tf.get_session_handle(one)
      x_handle = sess.run(tf.get_session_handle(one))

    # addition lives on GPU
    with tf.device("/gpu:0"):
      add_holder1, add_tensor1 = tf.get_session_tensor(dtype)
      add_holder2, add_tensor2 = tf.get_session_tensor(dtype)
      add_op = tf.add(add_tensor1, add_tensor2)
      add_output = tf.get_session_handle(add_op)


    # add 1 to tensor 20 times
    for i in range(20):
      x_handle = sess.run(add_output, feed_dict={add_holder1: one_handle.handle,
                                                 add_holder2: x_handle.handle})
```

The failure is
```

======================================================================
ERROR: testHandleDeletion (__main__.EnvTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/yaroslavvb/tfimmediate_macbook/tensorflow/bazel-bin/tensorflow/contrib/immediate/tensor_handle_test.runfiles/org_tensorflow/tensorflow/contrib/immediate/python/immediate/tensor_handle_test.py", line 35, in testHandleDeletion
    add_holder2: x_handle.handle})
  File "/Users/yaroslavvb/tfimmediate_macbook/tensorflow/bazel-bin/tensorflow/contrib/immediate/tensor_handle_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 333, in run
    run_metadata_ptr)
  File "/Users/yaroslavvb/tfimmediate_macbook/tensorflow/bazel-bin/tensorflow/contrib/immediate/tensor_handle_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 591, in _run
    self._register_dead_handle(handle)
  File "/Users/yaroslavvb/tfimmediate_macbook/tensorflow/bazel-bin/tensorflow/contrib/immediate/tensor_handle_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 726, in _register_dead_handle
    self.run(fetches, feed_dict=feeds)
  File "/Users/yaroslavvb/tfimmediate_macbook/tensorflow/bazel-bin/tensorflow/contrib/immediate/tensor_handle_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 333, in run
    run_metadata_ptr)
  File "/Users/yaroslavvb/tfimmediate_macbook/tensorflow/bazel-bin/tensorflow/contrib/immediate/tensor_handle_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 587, in _run
    feed_dict_string, options, run_metadata)
  File "/Users/yaroslavvb/tfimmediate_macbook/tensorflow/bazel-bin/tensorflow/contrib/immediate/tensor_handle_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 662, in _do_run
    target_list, options, run_metadata)
  File "/Users/yaroslavvb/tfimmediate_macbook/tensorflow/bazel-bin/tensorflow/contrib/immediate/tensor_handle_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 682, in _do_call
    raise type(e)(node_def, op, message)
InternalError: Unable to get element from the feed.

```

@yuanbyu @keveman

**Comment by yaroslavvb** [comment]
Actually for fix above to work, `_auto_gc_enabled` has to be set to `False` for all TensorHandles. The issue is that when `update_with_movers` moves CPU Tensor to GPU, it'll register CPU tensor handle for deletion. But if `_auto_gc_enabled` is set (it is set by default), then when Python runtime garbage collects that `TensorHandle` object, it'll register same handle for deletion again, so the program will fail on next gc pass with something like

`InvalidArgumentError: Failed to delete a tensor with handle 'GetSessionHandle_4;9;/job:localhost/replica:0/task:0/cpu:0' in the session store.
`
